src/sm2/field.rs

```diff
@@ -326,82 +339,160 @@ fn raw_mul(a: &FieldElem, b: &FieldElem) -> [u32; 16] {
     let mut carry: u64 = 0;
     let mut ret: [u32; 16] = [0; 16];
 
-    let mut k = 0;
-    while k < 15 {
-        let index = 15 - k;
-        let mut i = 0;
-        while i < 8 {
-            if i > k {
+    let mut ret_idx = 0;
+    while ret_idx < 15 {
+        let index = 15 - ret_idx;
+        let mut a_idx = 0;
+        while a_idx < 8 {
+            if a_idx > ret_idx {
                 break;
             }
-            let j = k - i;
-            if j < 8 {
-                let (u, v) = u32_mul(a.value[7 - i], b.value[7 - j]);
-                local += v;
-                carry += u;
+            let b_idx = ret_idx - a_idx;
+            if b_idx < 8 {
+                let (hi, lo) = u32_mul(a.value[7 - a_idx], b.value[7 - b_idx]);
+                local += lo;
+                carry += hi;
             }
 
-            i = i + 1;
+            a_idx += 1;
         }
         carry += local >> 32;
-        local = local & 0xffffffff;
+        local &= 0xffff_ffff;
         ret[index] = local as u32;
         local = carry;
         carry = 0;
 
-        k = k + 1;
+        ret_idx += 1;
     }
     ret[0] = local as u32;
     ret
 }
 
+impl ::std::cmp::PartialEq for FieldElem {
+    #[inline]
+    fn eq(&self, other: &Self) -> bool {
+        let lhs = self.value;
+        let rhs = other.value;
+        if lhs[0] != rhs[0] {
+            return false;
+        }
+        if lhs[1] != rhs[1] {
+            return false;
+        }
```

Comment:
```
lhs[0] != rhs[0] && lhs[1] != rhs[1] ... 
```

Reply:
OK, I will update this in the last commit then `push --force`.

I just kept the original logic of this piece of code, I did not change it.
I think changes for style and changes for logic should be in different commits.
https://github.com/cryptape/libsm/blob/3f62eb92758df72eb8f5deec71c9702b82d22eaf/src/sm2/field.rs#L395-L405